tpu/templates/tpu_grl_estimator/trainer.py

```diff
@@ -0,0 +1,308 @@
+    dataset = dataset.prefetch(tf.contrib.data.AUTOTUNE)
+
+    return dataset
+
+
+def main(args):
+    # pass the args as params so the model_fn can use
+    # the TPU specific args
+    params = vars(args)
+
+    if args.use_tpu:
+        # additional configs required for using TPUs
+        tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(args.tpu)
+        tpu_config = tf.contrib.tpu.TPUConfig(
+            num_shards=8, # using Cloud TPU v2-8
+            iterations_per_loop=args.save_checkpoints_steps)
+
+        # use the TPU version of RunConfig
+        config = tf.contrib.tpu.RunConfig(
+            cluster=tpu_cluster_resolver,
+            model_dir=args.model_dir,
+            tpu_config=tpu_config,
+            save_checkpoints_steps=args.save_checkpoints_steps,
+            save_summary_steps=100)
+
+        # TPUEstimator
+        estimator = tf.contrib.tpu.TPUEstimator(
+            model_fn=model_fn,
+            config=config,
+            params=params,
+            train_batch_size=args.train_batch_size,
+            eval_batch_size=32, # FIXME
+            export_to_tpu=False)
+    else:
+        config = tf.estimator.RunConfig(model_dir=args.model_dir)
+
+        estimator = tf.estimator.Estimator(
+            model_fn,
+            config=config,
+            params=params)
+
+    estimator.train(train_input_fn, max_steps=args.max_steps)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        '--model-dir',
+        type=str,
+        default='/tmp/tpu-template')
+    parser.add_argument(
+        '--max-steps',
+        type=int,
+        default=1000)
+    parser.add_argument(
+        '--train-batch-size',
+        type=int,
+        default=16)
+    parser.add_argument(
+        '--save-checkpoints-steps',
+        type=int,
+        default=100)
+    parser.add_argument(
+        '--use-tpu',
+        action='store_true')
+    parser.add_argument(
+        '--tpu',
+        default=None)
+
+    parser.add_argument(
+        '--gr-weight',
+        default=1.0,
+        help='The weight for gradient reversal.')
+    parser.add_argument(
+        '--lambda',
+        default=1.0,
+        help='The regularization factor.')
+
+    args, _ = parser.parse_known_args()
+
+    # colab.research.google.com specific
+    import sys
+    if 'google.colab' in sys.modules:
+        import json
+        import os
+
+        # TODO(user): change this
+        args.model_dir = 'gs://your-gcs-bucket'
+
+        # When connected to the TPU runtime
+        if 'COLAB_TPU_ADDR' in os.environ:
+            tpu_grpc = 'grpc://{}'.format(os.environ['COLAB_TPU_ADDR'])
+
+            args.tpu = tpu_grpc
+            args.use_tpu = True
+
+            # Upload credentials to the TPU
+            with tf.Session(tpu_grpc) as sess:
+                data = json.load(open('/content/adc.json'))
+                tf.contrib.cloud.configure_gcs(sess, credentials=data)
```

Comment:
Is there an alternative to tf.contrib.cloud.configure_gcs

Reply:
let me look into that as a follow up PR around the time when TF 2.0 is releases.